### PR TITLE
Undo redo

### PR DESCRIPTION
**Pull request by RichardNemcovic**
Feature UndoRedo refactored and tested

### CODE REVIEW

File: .github/workflows/build_test_pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+name: Build code base
```
Comment (do-nat) [inline .github/workflows/build_test_pipeline.yml:1]: there's already _two_ pipelines in the repository, please delete this file 😃

File: JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    private void prepare() {
+        map = new HashMap<>();
+        mouseEvent = mock(MouseEvent.class);
+        creationTool = new CreationTool(argumentList.svgBezierFigure, map);
+        creationTool.createdFigure = argumentList.svgBezierFigure;
+        creationTool.editor = argumentList.editor;
+    }
+
+    public WhenUndoActionIsPerformed undo_action_is_performed() {
+        List <Figure> list = argumentList.svgBezierFigure.getDrawing().getChildren();
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/draw/WhenUndoActionIsPerformed.java:31]: is not used anymore, can be deleted

File: JHotDraw/src/test/java/org/jhotdraw/samples/svg/figures/EllipseToolDrawAcceptanceTest.java [hunk 3]
```diff
@@ -19,4 +20,9 @@ public void EllipseToolDrawTest(){
         when().mouseReleased();
         then().figureCreated();
     }
+
+    @Override
+    public Scenario<GivenMousePressedAndDragged, WhenMouseIsReleased, ThenAnEllipseIsDrawn> getScenario() {
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/samples/svg/figures/EllipseToolDrawAcceptanceTest.java:25]: what is the purpose of this method?

File: JHotDraw/src/test/java/org/jhotdraw/samples/svg/gui/CanvasToolBarTest.java [hunk 4]
```diff
@@ -14,4 +15,9 @@ public void ChanceBackgroundColor(){
         when().chanceColorFieldToGreen();
         then().thenBackgroundColorIsGreen();
     }
+
+    @Override
+    public Scenario<GivenCanvasToolBar, WhenCanvasToolBar, ThenCanvasToolBar> getScenario() {
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/samples/svg/gui/CanvasToolBarTest.java:20]: what is the purpose?

File: JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @Test
+    public void undoOrRedoTest_null(){
+        UndoRedoManager.UndoRedoAction action = undoRedoManager.undoOrRedo(null);
+        assertEquals(action, null);
+    }
+
+    @Test
+    public void undoOrRedoTest_other(){
```
Comment (do-nat) [inline JHotDraw/src/test/java/org/jhotdraw/undo/UndoRedoManagerTest.java:46]: I think you should not catch an exception in a test. Look at this:
https://www.baeldung.com/junit-assert-exception

**Review by do-nat — CHANGES_REQUESTED** [review]
The tests are failing (has to be fixed) and the pipeline has to removed. Furthermore, I have some questions regarding some sections of the code 😅

File: JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java [hunk 6]
```diff
@@ -155,21 +155,32 @@ public void discardAllEdits() {
...
+
+
+
+
+
+
+
+
+
+
```
Comment (do-nat) [inline JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java:168]: can be removed

File: JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java [hunk 7]
```diff
@@ -243,58 +252,71 @@ private void updateActions() {
...
         }
+        return returnAction;
     }
-    
+
     public void addPropertyChangeListener(PropertyChangeListener listener) {
         propertySupport.addPropertyChangeListener(listener);
     }
+
+    //TODO remove code smell (Speculative Generality)
```
Comment (do-nat) [inline JHotDraw/src/main/java/org/jhotdraw/undo/UndoRedoManager.java:318]: commented code should be removed

**Review by do-nat — CHANGES_REQUESTED** [review]

**Comment by RichardNemcovic** [comment]
Worked on other branch and creating a new pull request

**Pull request closed without merging**